File: app/src/main/java/cn/guluwa/gulumusic/view/NextSongButton.java
package cn.guluwa.gulumusic.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

import cn.guluwa.gulumusic.R;
import cn.guluwa.gulumusic.utils.AppUtils;

/**
 * Created by guluwa on 2018/1/19.
 */

public class NextSongButton extends View {

    /**
     * 画笔
     */
    private Paint paint;

    /**
     * view颜色
     */
    private int color;

    /**
     * view线条width
     */
    private int width;

    /**
     * view宽度
     */
    private int mViewWidth;

    /**
     * view高度
     */
    private int mViewHeight;

    /**
     * view路径
     */
    private Path mViewPath;

    public NextSongButton(Context context) {
        this(context, null);
    }

    public NextSongButton(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public NextSongButton(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initPaint();
    }

    private void initPaint() {
        color = getResources().getColor(R.color.play_view_black);
        width = AppUtils.INSTANCE.dp2px(getContext(), 2);
        mViewWidth = mViewHeight = AppUtils.INSTANCE.dp2px(getContext(), 20);

        paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setStyle(Paint.Style.STROKE);
        paint.setColor(color);
        paint.setStrokeWidth(width);
        paint.setStrokeCap(Paint.Cap.ROUND);
        paint.setStrokeJoin(Paint.Join.ROUND);

        mViewPath = new Path();
        mViewPath.moveTo(width + getPaddingLeft(), width + getPaddingTop());
        mViewPath.lineTo(width + getPaddingLeft(), mViewHeight - width + getPaddingTop());
        mViewPath.lineTo(mViewWidth - width + getPaddingLeft(), mViewHeight / 2 + getPaddingTop());
        mViewPath.lineTo(width + getPaddingLeft(), width + getPaddingTop());
        mViewPath.moveTo(mViewWidth - width + getPaddingLeft(), width + getPaddingTop());
        mViewPath.lineTo(mViewWidth - width + getPaddingLeft(), mViewHeight - width + getPaddingTop());
    }


    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        setMeasuredDimension(
                mViewWidth + getPaddingLeft() + getPaddingRight(),
                mViewHeight + getPaddingTop() + getPaddingBottom());
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        canvas.drawPath(mViewPath, paint);
    }
}

File: app/src/main/java/cn/guluwa/gulumusic/view/NextSongButton.kt
package cn.guluwa.gulumusic.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.util.AttributeSet
import android.view.View

import cn.guluwa.gulumusic.R
import cn.guluwa.gulumusic.utils.AppUtils

/**
 * Created by guluwa on 2018/1/19.
 */

class NextSongButton @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : View(context, attrs, defStyleAttr) {

    /**
     * 画笔
     */
    private var paint: Paint? = null

    /**
     * view颜色
     */
    private var color: Int = 0

    /**
     * view线条width
     */
    private var viewWidth: Int = 0

    /**
     * view宽度
     */
    private var mViewWidth: Int = 0

    /**
     * view高度
     */
    private var mViewHeight: Int = 0

    /**
     * view路径
     */
    private var mViewPath: Path? = null

    init {
        initPaint()
    }

    private fun initPaint() {
        color = resources.getColor(R.color.play_view_black)
        viewWidth = AppUtils.dp2px(context, 2f)
        mViewHeight = AppUtils.dp2px(context, 20f)
        mViewWidth = mViewHeight

        paint = Paint(Paint.ANTI_ALIAS_FLAG)
        paint!!.style = Paint.Style.STROKE
        paint!!.color = color
        paint!!.strokeWidth = viewWidth.toFloat()
        paint!!.strokeCap = Paint.Cap.ROUND
        paint!!.strokeJoin = Paint.Join.ROUND

        mViewPath = Path()
        mViewPath!!.moveTo((viewWidth + paddingLeft).toFloat(), (viewWidth + paddingTop).toFloat())
        mViewPath!!.lineTo((viewWidth + paddingLeft).toFloat(), (mViewHeight - viewWidth + paddingTop).toFloat())
        mViewPath!!.lineTo((mViewWidth - viewWidth + paddingLeft).toFloat(), (mViewHeight / 2 + paddingTop).toFloat())
        mViewPath!!.lineTo((viewWidth + paddingLeft).toFloat(), (viewWidth + paddingTop).toFloat())
        mViewPath!!.moveTo((mViewWidth - viewWidth + paddingLeft).toFloat(), (viewWidth + paddingTop).toFloat())
        mViewPath!!.lineTo((mViewWidth - viewWidth + paddingLeft).toFloat(), (mViewHeight - viewWidth + paddingTop).toFloat())
    }


    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        setMeasuredDimension(
                mViewWidth + paddingLeft + paddingRight,
                mViewHeight + paddingTop + paddingBottom)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        canvas.drawPath(mViewPath!!, paint!!)
    }
}
